File: app/src/main/java/com/example/workoutapplication/Exercise.java
package com.example.workoutapplication;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.LinkedList;
import java.util.UUID;

public class Exercise {

    // TODO: check uniqueness (unique name/description?)

    private UUID exerciseID; // Class with very low-probably of generating the same "unique ID"
    private String name;
    private String description;
    private String thumbnailID; // Passing the storing/loading of images onto the activity
    // private Workouts[] includedInWorkouts;
    private LinkedList<String> tags; // maybe switch this to pairs or a Tag class (tag name, reference to Tag)

    /**
     * Exercise constructor which takes a JsonString, in the format of .toJsonString()
     * @param JsonString
     * @return an Exercise with attributes based on JsonString
     */
    public static Exercise fromJsonString(String JsonString)
    {
        Exercise resultExercise = new Exercise();

        try
        {
            JSONObject jsonObject = new JSONObject(JsonString);

            String IDString = (String) jsonObject.get("UniqueID");
            resultExercise.setExerciseID(UUID.fromString(IDString));

            resultExercise.setName(jsonObject.get("Name").toString());
            resultExercise.setDescription(jsonObject.get("Description").toString());
            resultExercise.setThumbnailID(jsonObject.get("ThumbnailID").toString());

            JSONArray tagsArray = new JSONArray(jsonObject.get("Tags").toString());
            for (int i=0; i<tagsArray.length(); i++)
            {
                resultExercise.addTag(tagsArray.get(i).toString());
            }
        }
        catch (JSONException e)
        {
            e.printStackTrace();
        }

        return resultExercise;
    }

    private Exercise() {
        tags = new LinkedList<String>();
    }

    // Constructor for the creation of "new" exercises, with a name and description as the base
    public Exercise (String name, String description)
    {
        exerciseID = UUID.randomUUID();

        this.name = name;
        this.description = description;
        tags = new LinkedList<String>();
    }

    public Exercise (String uniqueID, String name, String description)
    {
        this.exerciseID = UUID.fromString(uniqueID);
        this.name = name;
        this.description = description;
        tags = new LinkedList<String>();
    }

    // Private setter, intended for use in .fromJsonString() method
    private void setExerciseID(UUID exerciseID)
    {
        this.exerciseID = exerciseID;
    }

    public UUID getExerciseID()
    {
        return this.exerciseID;
    }

    public String getName()
    {
        return name;
    }

    public void setName(String name)
    {
        this.name = name;
    }

    public String getDescription()
    {
        return description;
    }

    public void setDescription(String description)
    {
        this.description = description;
    }

    public String getThumbnailID()
    {
        return thumbnailID;
    }

    public void setThumbnailID(String thumbnailID)
    {
        this.thumbnailID = thumbnailID;
    }

    // Tag Methods
    public LinkedList<String> getTags()
    {
        return tags;
    }

    public void clearAllTags() {
        tags.clear();
    }

    public void removeTag(String tag)
    {
        tags.remove(tag);
    }

    public void addTag(String tag)
    {
        tags.add(tag);
    }

    // JSON Methods

    // Returns a Json string representation of the current object
    public String toJsonString()
    {
        // TODO: for boiler-plate reasons, maybe try and remove JSONObject and manually convert to strings --> no dependence on org.json
        JSONObject jsonExercise = new JSONObject();

        try
        {
            // Putting exercise's variable values
            jsonExercise.put("UniqueID", this.getExerciseID().toString());
            jsonExercise.put("Name", this.getName());
            jsonExercise.put("Description", this.getDescription());
            jsonExercise.put("ThumbnailID", this.getThumbnailID());

            // JSONArray to hold exercise's tags
            JSONArray array = new JSONArray(this.getTags());
            jsonExercise.put("Tags", array);
        }
        catch (JSONException e)
        {
            e.printStackTrace();
            return null; /** INCREDIBLY IMPORTANT **/
        }

        return jsonExercise.toString();
    }
}

File: app/src/main/java/com/example/workoutapplication/Exercise.kt
package com.example.workoutapplication

import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.util.LinkedList
import java.util.UUID

class Exercise {
    // TODO: check uniqueness (unique name/description?)
    var exerciseID: UUID? = null // Class with very low-probably of generating the same "unique ID"
        private set
    @JvmField
    var name: String? = null
    @JvmField
    var description: String? = null
    @JvmField
    var thumbnailID: String? = null // Passing the storing/loading of images onto the activity

    // Tag Methods
    var tags // maybe switch this to pairs or a Tag class (tag name, reference to Tag)
            : LinkedList<String?>
        private set

    private constructor() {
        tags = LinkedList()
    }

    // Constructor for the creation of "new" exercises, with a name and description as the base
    constructor(name: String?, description: String?) {
        exerciseID = UUID.randomUUID()
        this.name = name
        this.description = description
        tags = LinkedList()
    }

    constructor(uniqueID: String?, name: String?, description: String?) {
        exerciseID = UUID.fromString(uniqueID)
        this.name = name
        this.description = description
        tags = LinkedList()
    }

    // Private setter, intended for use in .fromJsonString() method
    private fun setExerciseID(exerciseID: UUID) {
        this.exerciseID = exerciseID
    }

    fun clearAllTags() {
        tags.clear()
    }

    fun removeTag(tag: String?) {
        tags.remove(tag)
    }

    fun addTag(tag: String?) {
        tags.add(tag)
    }

    // JSON Methods
    // Returns a Json string representation of the current object
    fun toJsonString(): String {
        // TODO: for boiler-plate reasons, maybe try and remove JSONObject and manually convert to strings --> no dependence on org.json
        val jsonExercise = JSONObject()
        try {
            // Putting exercise's variable values
            jsonExercise.put("UniqueID", exerciseID.toString())
            jsonExercise.put("Name", name)
            jsonExercise.put("Description", description)
            jsonExercise.put("ThumbnailID", thumbnailID)

            // JSONArray to hold exercise's tags
            val array = JSONArray(tags)
            jsonExercise.put("Tags", array)
        } catch (e: JSONException) {
            e.printStackTrace()
            return "Json conversion error for $exerciseID"
        }
        return jsonExercise.toString()
    }

    companion object {
        /**
         * Exercise constructor which takes a JsonString, in the format of .toJsonString()
         * @param JsonString
         * @return an Exercise with attributes based on JsonString
         */
        fun fromJsonString(jsonString: String?): Exercise {
            val resultExercise = Exercise()
            try {
                val jsonObject = JSONObject(jsonString)
                resultExercise.apply() {
                    val IDString = jsonObject["UniqueID"] as String
                    exerciseID = UUID.fromString(IDString)
                    name = jsonObject["Name"].toString()
                    description = jsonObject["Description"].toString()
                    thumbnailID = jsonObject["ThumbnailID"].toString()

                    val tagsArray = JSONArray(jsonObject["Tags"].toString())
                    for (i in 0 until tagsArray.length()) {
                        resultExercise.addTag(tagsArray[i].toString())
                    }
                }
            } catch (e: JSONException) {
                e.printStackTrace()
            }
            return resultExercise
        }
    }
}